gio/src/subclass/async_result.rs

```diff
@@ -0,0 +1,108 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use glib::object::Cast;
+use glib::translate::*;
+
+use glib::subclass::prelude::*;
+
+use crate::AsyncResult;
+
+pub trait AsyncResultImpl: ObjectImpl {
+    fn user_data(&self, res: &Self::Type) -> glib::Pointer;
```

Comment:
Do you see any way of making this less unsafe?

Reply:
No, but Rust programs will probably never want to call that, it's only for compatibility with C FFI

Reply:
I wonder why the user_data/source_object/tag handling is not just done directly in the base class

Reply:
Because it is an interface, not a class :) The only thing I marked unsafe is the constructor, that should be appropriate because the C callback is expected to deref the pointer? Or should the getter also be unsafe? I don't think there is any reason to deref the pointer from rust

Reply:
We could in theory do the handling of user_data/source_object/tag as part of the [`TypeData`'s `class_data`](https://gtk-rs.org/gtk-rs-core/git/docs/glib/subclass/types/struct.TypeData.html#method.set_class_data)? Unless I'm missing something there is exactly one valid implementation for these so we can implement that ourselves?

Reply:
I am going to profile this again to figure out why exactly it's faster, I would actually prefer to get rid of this and use `GTask`